contracts/BondCalculator.sol

```diff
@@ -0,0 +1,301 @@
+            if (vPower >= localParams.targetVotingPower) {
+                discountBooster += uint256(localParams.weightFactors[3]) * 1e18;
+            } else {
+                discountBooster += (uint256(localParams.weightFactors[3]) * vPower * 1e18) /
+                    uint256(localParams.targetVotingPower);
+            }
+        }
+
+        // Normalize discount booster by the max sum of weights
+        discountBooster /= MAX_SUM_WEIGHTS;
+
+        // IDF = 1 + normalized booster
+        idf = 1e18 + discountBooster;
+    }
+
+    /// @dev Calculates the amount of OLAS tokens based on the bonding calculator mechanism accounting for dynamic IDF.
+    /// @param account Account address.
+    /// @param tokenAmount LP token amount.
+    /// @param priceLP LP token price.
+    /// @param bondVestingTime Bond vesting time.
+    /// @param productMaxVestingTime Product max vesting time.
+    /// @param productSupply Current product supply.
+    /// @param productPayout Current product payout.
+    /// @return amountOLAS Resulting amount of OLAS tokens.
+    function calculatePayoutOLAS(
+        address account,
+        uint256 tokenAmount,
+        uint256 priceLP,
+        uint256 bondVestingTime,
+        uint256 productMaxVestingTime,
+        uint256 productSupply,
+        uint256 productPayout
+    ) external view returns (uint256 amountOLAS) {
+        // The result is divided by additional 1e18, since it was multiplied by in the current LP price calculation
+        // The resulting amountDF can not overflow by the following calculations: idf = 64 bits;
+        // priceLP = 2 * r0/L * 10^18 = 2*r0*10^18/sqrt(r0*r1) ~= 61 + 96 - sqrt(96 * 112) ~= 53 bits (if LP is balanced)
+        // or 2* r0/sqrt(r0) * 10^18 => 87 bits + 60 bits = 147 bits (if LP is unbalanced);
+        // tokenAmount is of the order of sqrt(r0*r1) ~ 104 bits (if balanced) or sqrt(96) ~ 10 bits (if max unbalanced);
+        // overall: 64 + 53 + 104 = 221 < 256 - regular case if LP is balanced, and 64 + 147 + 10 = 221 < 256 if unbalanced
+        // mulDiv will correctly fit the total amount up to the value of max uint256, i.e., max of priceLP and max of tokenAmount,
+        // however their multiplication can not be bigger than the max of uint192
+        uint256 totalTokenValue = mulDiv(priceLP, tokenAmount, 1);
+        // Check for the cumulative LP tokens value limit
+        if (totalTokenValue > type(uint192).max) {
+            revert Overflow(totalTokenValue, type(uint192).max);
+        }
+
+        // Calculate the dynamic inverse discount factor
+        uint256 idf = calculateIDF(account, bondVestingTime, productMaxVestingTime, productSupply, productPayout);
+
+        // Amount with the discount factor is IDF * priceLP * tokenAmount / 1e36
+        // At this point of time IDF is bound by the max of uint64, and totalTokenValue is no bigger than the max of uint192
+        amountOLAS = (idf * totalTokenValue) / 1e36;
```

Comment:
This is the only new difference - tokenomics-calculated IDF is changed with the local calculated dynamic IDF according to the proposed approach.